### DISCUSSION TITLE
Make mirror mod in osu!standard affect stacking direction

### DISCUSSION THREAD

**Discussion by cl8n** [ppy/osu#29327] [Ideas]
(was prompted by #29324 to open this because it feels to me like something that should be considered prior to ranked)

currently, the Mirror mod reflects positions of hit objects, but it *doesn't* change the stacking direction, which still goes to the top-left. 

this takes away from the purpose of Mirror imo, which should be to give you an identical play experience, just reflected on one or both axes. it feels like a random/obscure technicality to have that expectation broken, and leaves this mod in a weird spot of being almost "just preference" except not really because the map does change.

the argument I saw in favor of the current behavior is that it matches HR. I don't think the behavior of a mod with an unrelated purpose should be influencing how new mods work, though, especially when it is to the detriment of the new mod

### about supporting existing scores

I didn't originally mention this because I thought I wouldn't need to, but I don't mean to imply that old scores using Mirror should be wiped. if Mirror were to be updated with my suggestion here, then I think it should be done in a way that the old behavior can still be selectively applied for old scores. my idea for an implementation is adding a new setting to Mirror that exposes no actual control to the user, effectively making a "version 2" of the mod -- but I'll leave the details to whoever would actually code this.

**Comment by Givikap120** [comment]
Wiping all MR scores over something like this is just silly
The minmax effect is practically not real because the difference in 0.0001 star is not minmaxable

**Reply by cl8n** [reply]
I did not imply wiping scores
as for "minmax" thing yeah I'll just edit that out, it's not my main point anyway

**Reply by Givikap120** [reply]
If you're going to change the behaviour of the mod - you need to wipe all scores set with this mod

**Reply by cl8n** [reply]
updated again to clarify what I was thinking w.r.t old scores

**Comment by minisbett** [comment]
This won't work without wiping all scores so it's not going to happen, as much as I'd like it to

**Reply by cl8n** [reply]
I updated OP to explain what I was thinking w.r.t old scores (I didn't mean to imply they should be wiped)

**Comment by Givikap120** [comment]
Version 2 gonna be very messy and peppy/bdach will never allow something like this
Adding new setting is much more likely, but it's still the 0.0001% probability vs 1% probability
99% probability is still that you will need to wipe scores

**Comment by peppy** [comment]
Scores don't really matter that much for new mods with changes like this IMO.

**Reply by minisbett** [reply]
You think it'd be fine to just update stacks to change direction on mirror?

As for old scores, I guess the chance of them being affected because of a big stack and unfortunate aim is slim, + the mod is quite new so I think yeah if anything now would be the time to change it.

**Reply by peppy** [reply]
There are 41,961 scores set using mirror mod. Of those, 2,063 are S/SS rank.

**Reply by Tenexxt** [reply]
_**WARNING** 
This reply could be absolutely stupid because I'm not 100% sure how mods and overall system works._

If there would be problems because of the stack direction, the current ver of mirror could be made unavailable and any score that has problems because of the way the new mirror would work, you could assign it manually as a (old) toggle for mirror, which would make the mirror work the way it works now. The toggle could be available as a option for the mod, or hidden to only be applied to scores manually by devs if the score is broken/it was requested by the player who set the score

**Reply by cl8n** [reply]
I made a quick mockup for something like the above a few days ago <https://github.com/ppy/osu/commit/7e1df8dccef53ed7a0778c7bd8aa125c31da96f1> (this doesn't include the actual mirror behavior). the gist is that mods could have a "version", which defaults to latest when instantiating the mod, but defaults to 0 when deserializing an APIMod with no version info. it's not a lot of lines of code but I can't tell if the concept would be considered unnecessarily complicated.

**Reply by peppy** [reply]
we're not versioning mods, i'd sooner wip escores.

**Reply by Tenexxt** [reply]
Versioning scores would become very annoying to not only to devs but to players also, because you'd have to remember all versions to know what is what and the amount of space over time it would take would be annoying to sort out in the code, and the amount of bugs that could come just because of that is annoyingly high.

That why I'd say that if we were to change the way mirror works, it's better to add not mirroring stacking as a trigger, apply it to every score that uses mirror, so there will be pretty much no way someone looses scores.
Also this is definitely a time to apply this change because adding changes to mods is very discouraged IMO, exept few factors:

- Lazer only mods: Lazer mods are new and developed still. Of course this depends on the overall amount of scores on a mod, but the overall number will still be lower than mods that were also in stable.
- Multiplier changes: since only score multiplier is changed, all that has to be done is run a recalculation on all scores with the changed multiplier

Other than that, I think it's preferred to not change  mods to a higher part, unless it's a trigger that changes some of its parts

**Reply by cl8n** [reply]
> we're not versioning mods, i'd sooner wip escores.

I'm not gonna pretend like there are enough MR scores to warrant this but I do think it would leave "significant" ones in an odd spot, for example those in daily challenge, I don't think those should ever be wiped bc ppl want to look back on accurate records/stats there

**Reply by Tenexxt** [reply]
> I don't think those should ever be wiped bc ppl want to look back on accurate records/stats there

Also because the scores would be gone, depending on how the system works (I'm not deep enough to know that) it could break peoples streaks, because if the system checks for the score, and the score is gone, and if a recalculation is run for whatever reason - the streak system will be in shambles, especially if it were to happen to people with 3+ months streaks

**Comment by TextAdventurer12** [comment]
maybe this could become a toggle in the mirror mod, since that would remove any need to wipe anything and people who prefer the current stacking would still be satisfied

**Reply by minisbett** [reply]
A "version 2" was already mentioned but I don't think the direction of stacks is interesting enough to the user to become an option, and adding an option for the sake of backwards compatibility idk about that

**Comment by NoelleTGS** [comment]
Personally this would really turn me off from the mod, I (and I would assume most other people who use it) use MR solely for mindblock purposes and this would really screw with my reading and muscle memory. It's essentially the only negative part of using the monitor flipping method that stable players use.

**Reply by cl8n** [reply]
well I can't say that's relatable for me as someone who used to flip monitor too. you're able to read the rest of the map flipped just fine and only have issues with this specific pattern? do you also have trouble reading manual stacks in different directions?

**Reply by NoelleTGS** [reply]
In all honesty I could just be thinking of something different, I just remember it being the one thing that felt unnatural when I would try flipping monitor. Might need to test again to see if I can narrow down what bothered me about it.

**Reply by Natelytle** [reply]
I agree with Noelle here, I think I would use mirror way less if stacks changed direction. It just looks and feels really ugly.
